// @progress/kendo-vue-pdf/dist/es/treelist/TreeListPDFExport.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version && allVue.version[0] === '3';
import { savePDF } from '../savePDF';
import { provideSaveTreeListPDF } from './provideSaveTreeListPDF';
import { getDefaultSlots } from '@progress/kendo-vue-common';
/**
 * @hidden
 */
var TreeListPDFExportVue2 = {
  name: 'treelist-pdf-export',
  props: {
    author: String,
    avoidLinks: [Boolean, String],
    forcePageBreak: String,
    keepTogether: String,
    creator: String,
    date: Date,
    imageResolution: Number,
    fileName: String,
    forceProxy: Boolean,
    keywords: String,
    landscape: Boolean,
    margin: [String, Object],
    pageTemplate: [Object, Function],
    paperSize: [String, Object],
    repeatHeaders: Boolean,
    scale: Number,
    proxyData: Object,
    proxyURL: String,
    proxyTarget: String,
    producer: String,
    subject: String,
    title: String
  },
  created: function created() {
    this.saveTreeListPDF = provideSaveTreeListPDF(this.getSavePDF());
  },
  methods: {
    save: function save(data, callback) {
      var defaultSlot = getDefaultSlots(this);
      // @ts-ignore
      var treeListSlot = defaultSlot.filter(function (child) {
        return child && child.tag && child.tag.toLowerCase().indexOf('treelist') !== -1 || child.componentOptions && child.componentOptions.tag && child.componentOptions.tag.toLowerCase().indexOf('treelist') !== -1 || child.type && child.type.name && child.type.name.toLowerCase().indexOf('treelist') !== -1;
      });
      if (treeListSlot.length) {
        this.treeListComponent = this.v3 ? treeListSlot[0] : treeListSlot[0].componentInstance;
        this.saveTreeListPDF(this.treeListComponent, this.$props, callback, data, this.v3);
      }
    },
    getSavePDF: function getSavePDF() {
      return savePDF;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlot = getDefaultSlots(this);
    return h("div", [defaultSlot]);
  }
};
/**
 * @hidden
 */
var TreeListPDFExport = TreeListPDFExportVue2;
export { TreeListPDFExport, TreeListPDFExportVue2 };
