// @progress/kendo-vue-common/dist/npm/main.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.shouldShowValidationUI = exports.validatePackage = exports.setter = exports.getter = exports.getDir = exports.isRtl = exports.getDefaultSlots = exports.getTabIndex = exports.templateDefinition = exports.templateRendering = exports.getTemplate = exports.cloneDate = exports.cloneObject = exports.cloneValue = exports.cloneArray = exports.Draggable = exports.clone = exports.focusContainer = exports.focusLastFocusableChild = exports.lastFocusableChild = exports.focusFirstFocusableChild = exports.firstFocusableChild = exports.canUseDOM = exports.hasListener = exports.getListeners = exports.isObject = exports.noop = exports.Keys = exports.guid = exports.classNames = exports.setRef = exports.getRef = void 0;
var classNames_1 = require("./classNames");
Object.defineProperty(exports, "classNames", { enumerable: true, get: function () { return classNames_1.classNames; } });
var guid_1 = require("./guid");
Object.defineProperty(exports, "guid", { enumerable: true, get: function () { return guid_1.guid; } });
var keys_1 = require("./keys");
Object.defineProperty(exports, "Keys", { enumerable: true, get: function () { return keys_1.Keys; } });
var canUseDOM_1 = require("./canUseDOM");
Object.defineProperty(exports, "canUseDOM", { enumerable: true, get: function () { return canUseDOM_1.canUseDOM; } });
var refs_1 = require("./refs");
Object.defineProperty(exports, "getRef", { enumerable: true, get: function () { return refs_1.getRef; } });
Object.defineProperty(exports, "setRef", { enumerable: true, get: function () { return refs_1.setRef; } });
var focus_1 = require("./focus");
Object.defineProperty(exports, "focusFirstFocusableChild", { enumerable: true, get: function () { return focus_1.focusFirstFocusableChild; } });
Object.defineProperty(exports, "focusLastFocusableChild", { enumerable: true, get: function () { return focus_1.focusLastFocusableChild; } });
Object.defineProperty(exports, "focusContainer", { enumerable: true, get: function () { return focus_1.focusContainer; } });
Object.defineProperty(exports, "firstFocusableChild", { enumerable: true, get: function () { return focus_1.firstFocusableChild; } });
Object.defineProperty(exports, "lastFocusableChild", { enumerable: true, get: function () { return focus_1.lastFocusableChild; } });
var Draggable_1 = require("./Draggable");
Object.defineProperty(exports, "Draggable", { enumerable: true, get: function () { return Draggable_1.Draggable; } });
var clone_1 = require("./clone");
Object.defineProperty(exports, "clone", { enumerable: true, get: function () { return clone_1.clone; } });
Object.defineProperty(exports, "cloneArray", { enumerable: true, get: function () { return clone_1.cloneArray; } });
Object.defineProperty(exports, "cloneValue", { enumerable: true, get: function () { return clone_1.cloneValue; } });
Object.defineProperty(exports, "cloneObject", { enumerable: true, get: function () { return clone_1.cloneObject; } });
Object.defineProperty(exports, "cloneDate", { enumerable: true, get: function () { return clone_1.cloneDate; } });
var isObject_1 = require("./isObject");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return isObject_1.isObject; } });
var noop_1 = require("./noop");
Object.defineProperty(exports, "noop", { enumerable: true, get: function () { return noop_1.noop; } });
var getter_1 = require("./getter");
Object.defineProperty(exports, "getter", { enumerable: true, get: function () { return getter_1.getter; } });
var setter_1 = require("./setter");
Object.defineProperty(exports, "setter", { enumerable: true, get: function () { return setter_1.setter; } });
var listeners_1 = require("./listeners");
Object.defineProperty(exports, "hasListener", { enumerable: true, get: function () { return listeners_1.hasListener; } });
Object.defineProperty(exports, "getListeners", { enumerable: true, get: function () { return listeners_1.getListeners; } });
var templateRendering_1 = require("./templateRendering");
Object.defineProperty(exports, "templateRendering", { enumerable: true, get: function () { return templateRendering_1.templateRendering; } });
Object.defineProperty(exports, "getTemplate", { enumerable: true, get: function () { return templateRendering_1.getTemplate; } });
Object.defineProperty(exports, "templateDefinition", { enumerable: true, get: function () { return templateRendering_1.templateDefinition; } });
var getTabIndex_1 = require("./getTabIndex");
Object.defineProperty(exports, "getTabIndex", { enumerable: true, get: function () { return getTabIndex_1.getTabIndex; } });
var defaultSlots_1 = require("./defaultSlots");
Object.defineProperty(exports, "getDefaultSlots", { enumerable: true, get: function () { return defaultSlots_1.getDefaultSlots; } });
var isRtl_1 = require("./isRtl");
Object.defineProperty(exports, "isRtl", { enumerable: true, get: function () { return isRtl_1.isRtl; } });
Object.defineProperty(exports, "getDir", { enumerable: true, get: function () { return isRtl_1.getDir; } });
var validate_package_1 = require("./validate-package");
Object.defineProperty(exports, "validatePackage", { enumerable: true, get: function () { return validate_package_1.validatePackage; } });
Object.defineProperty(exports, "shouldShowValidationUI", { enumerable: true, get: function () { return validate_package_1.shouldShowValidationUI; } });
__exportStar(require("./theme"), exports);
__exportStar(require("./treeDataOperations"), exports);
__exportStar(require("./browser-support.service"), exports);
__exportStar(require("./scrollbarWidth"), exports);
__exportStar(require("./hasRelativeStackingContext"), exports);
__exportStar(require("./icons/Icon"), exports);
__exportStar(require("./icons/FontIcon"), exports);
__exportStar(require("./icons/SvgIcon"), exports);
__exportStar(require("./icons/getIconName"), exports);
__exportStar(require("./icons/models/flip"), exports);
__exportStar(require("./icons/models/size"), exports);
__exportStar(require("./icons/models/theme-color"), exports);
__exportStar(require("./navigation"), exports);
__exportStar(require("./getActiveElement"), exports);
__exportStar(require("./watermark/WatermarkOverlay"), exports);
