// @progress/kendo-vue-common/dist/esm/main.js
import { classNames } from './classNames.js';
import { guid } from './guid.js';
import { Keys } from './keys.js';
import { canUseDOM } from './canUseDOM.js';
import { getRef, setRef } from './refs.js';
import { focusFirstFocusableChild, focusLastFocusableChild, focusContainer, firstFocusableChild, lastFocusableChild } from './focus.js';
import { Draggable } from './Draggable.js';
import { clone, cloneArray, cloneValue, cloneObject, cloneDate } from './clone.js';
import { isObject } from './isObject.js';
import { noop } from './noop.js';
import { getter } from './getter.js';
import { setter } from './setter.js';
import { hasListener, getListeners } from './listeners.js';
import { templateRendering, getTemplate, templateDefinition } from './templateRendering.js';
import { getTabIndex } from './getTabIndex.js';
import { getDefaultSlots } from './defaultSlots.js';
import { isRtl, getDir } from './isRtl.js';
import { validatePackage, shouldShowValidationUI } from './validate-package.js';
export { getRef, setRef, classNames, guid, Keys, noop, isObject, getListeners, hasListener, canUseDOM, firstFocusableChild, focusFirstFocusableChild, lastFocusableChild, focusLastFocusableChild, focusContainer, clone, Draggable, cloneArray, cloneValue, cloneObject, cloneDate, getTemplate, templateRendering, templateDefinition, getTabIndex, getDefaultSlots, isRtl, getDir, getter, setter, validatePackage, shouldShowValidationUI };
export * from './theme.js';
export * from './treeDataOperations.js';
export * from './browser-support.service.js';
export * from './scrollbarWidth.js';
export * from './hasRelativeStackingContext.js';
export * from './icons/Icon.js';
export * from './icons/FontIcon.js';
export * from './icons/SvgIcon.js';
export * from './icons/getIconName.js';
export * from './icons/models/flip.js';
export * from './icons/models/size.js';
export * from './icons/models/theme-color.js';
export * from './navigation.js';
export * from './getActiveElement.js';
export * from './watermark/WatermarkOverlay.js';
