// @progress/kendo-vue-pdf/dist/npm/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreeListPDFExportVue2 = exports.TreeListPDFExport = exports.GridPdfExportVue2 = exports.GridPdfExport = exports.savePDF = exports.PDFExportVue2 = exports.PDFExport = void 0;
var PDFExport_1 = require("./PDFExport");
Object.defineProperty(exports, "PDFExport", { enumerable: true, get: function () { return PDFExport_1.PDFExport; } });
Object.defineProperty(exports, "PDFExportVue2", { enumerable: true, get: function () { return PDFExport_1.PDFExportVue2; } });
var savePDF_1 = require("./savePDF");
Object.defineProperty(exports, "savePDF", { enumerable: true, get: function () { return savePDF_1.savePDF; } });
var GridPDFExport_1 = require("./grid/GridPDFExport");
Object.defineProperty(exports, "GridPdfExport", { enumerable: true, get: function () { return GridPDFExport_1.GridPdfExport; } });
Object.defineProperty(exports, "GridPdfExportVue2", { enumerable: true, get: function () { return GridPDFExport_1.GridPdfExportVue2; } });
var TreeListPDFExport_1 = require("./treelist/TreeListPDFExport");
Object.defineProperty(exports, "TreeListPDFExport", { enumerable: true, get: function () { return TreeListPDFExport_1.TreeListPDFExport; } });
Object.defineProperty(exports, "TreeListPDFExportVue2", { enumerable: true, get: function () { return TreeListPDFExport_1.TreeListPDFExportVue2; } });
// Automatic installation if Vue has been added to the global scope.
var vue = 'Vue';
if (typeof window !== 'undefined' && window[vue] && window[vue].component) {
    window[vue].component('kendo-pdf-export', PDFExport_1.PDFExport);
    window[vue].component('kendo-grid-pdf-export', GridPDFExport_1.GridPdfExport);
}
